### PR TITLE
Resolve ILLink warnings in Microsoft.VisualBasic.Core

### PR DESCRIPTION
**Pull request by joperezr**
cc: @eerhardt @vitek-karas @agocke 

Addressing all Linker warnings on Microsoft.VisualBasic.Core assembly.

Contributes to https://github.com/dotnet/runtime/issues/45623

### CODE REVIEW

**Comment by agocke** [comment]
@cston FYI

File: src/libraries/Microsoft.VisualBasic.Core/ref/Microsoft.VisualBasic.Core.cs [hunk 1]
```diff
@@ -197,7 +197,9 @@ public sealed partial class ControlChars
     public sealed partial class Conversion
     {
         internal Conversion() { }
-        public static object CTypeDynamic(object? Expression, System.Type TargetType) { throw null; }
+        [System.Diagnostics.CodeAnalysis.RequiresUnreferencedCodeAttribute("The Expression origin object cannot be statically analyzed and may be trimmed.")]
```
Comment (eerhardt) [inline src/libraries/Microsoft.VisualBasic.Core/ref/Microsoft.VisualBasic.Core.cs:200]: (nit) it isn't really the object that might be trimmed, but instead the members it points to. What do you think of:

`The Expression's underlying type cannot be statically analyzed and its members may be trimmed.`

File: src/libraries/Microsoft.VisualBasic.Core/ref/Microsoft.VisualBasic.Core.cs [hunk 2]
```diff
@@ -500,9 +509,11 @@ public sealed partial class Interaction
         public static void AppActivate(string Title) { }
         [System.Runtime.Versioning.SupportedOSPlatformAttribute("windows")]
         public static void Beep() { }
+        [System.Diagnostics.CodeAnalysis.RequiresUnreferencedCodeAttribute("The procedure name cannot be statically analyzed and may be trimmed.")]
```
Comment (eerhardt) [inline src/libraries/Microsoft.VisualBasic.Core/ref/Microsoft.VisualBasic.Core.cs:512]: It's not the procedure name that can't be statically analyzed, right? It is the type of `ObjectRef`.

File: src/libraries/Microsoft.VisualBasic.Core/ref/Microsoft.VisualBasic.Core.cs [hunk 3]
```diff
@@ -631,6 +643,7 @@ public sealed partial class Strings
         public static int Len(short Expression) { throw null; }
         public static int Len(int Expression) { throw null; }
         public static int Len(long Expression) { throw null; }
+        [System.Diagnostics.CodeAnalysis.RequiresUnreferencedCodeAttribute("The object in the expression cannot be statically analyzed and may be trimmed")]
```
Comment (eerhardt) [inline src/libraries/Microsoft.VisualBasic.Core/ref/Microsoft.VisualBasic.Core.cs:646]: Similar to above - it isn't the object, but the object's type.

File: src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/ConversionResolution.vb [hunk 4]
```diff
@@ -114,12 +115,19 @@ Namespace Microsoft.VisualBasic.CompilerServices
...
 #If DEBUG Then
-            VerifyForLoopWidestType()
+            VerifyForLoopWidestTypeInDebugMode()
 #End If
         End Sub
-
+        
 #If DEBUG Then
+        <UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode",
+            Justification:="This method only gets called in DEBUG mode.")>
```
Comment (eerhardt) [inline src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/ConversionResolution.vb:124]: Isn't it always? Since it is declared inside `#If DEBUG`?

File: src/libraries/Microsoft.VisualBasic.Core/ref/Microsoft.VisualBasic.Core.cs [hunk 5]
```diff
@@ -890,30 +905,38 @@ public sealed partial class LateBinding
         internal LateBinding() { }
         [System.Diagnostics.DebuggerHiddenAttribute]
         [System.Diagnostics.DebuggerStepThroughAttribute]
-        public static void LateCall(object? o, System.Type? objType, string name, object?[]? args, string?[]? paramnames, bool[]? CopyBack) { }
+        [System.Diagnostics.CodeAnalysis.RequiresUnreferencedCodeAttribute("Late binding is dynamic and cannot be statically analyzed, so it may be trimmed")]
```
Comment (eerhardt) [inline src/libraries/Microsoft.VisualBasic.Core/ref/Microsoft.VisualBasic.Core.cs:908]: > so it may be trimmed

This could be a little more precise.  "Late binding is dynamic and cannot be statically analyzed. The referenced types and members may be trimmed"

File: src/libraries/Microsoft.VisualBasic.Core/src/MatchingRefApiCompatBaseline.txt [hunk 6]
```diff
@@ -1,28 +1,4 @@
-Compat issues with assembly Microsoft.VisualBasic:
-
-# Loaded via reflection so needs to be public in the implementation but we don't want to expose them publicly in the ref assembly:
```
Comment (eerhardt) [inline src/libraries/Microsoft.VisualBasic.Core/src/MatchingRefApiCompatBaseline.txt:3]: Are these changes intentional? What is the reasoning?

File: src/libraries/Microsoft.VisualBasic.Core/src/MatchingRefApiCompatBaseline.txt [hunk 7]
```diff
@@ -1,28 +1,4 @@
...
-TypesMustExist : Type 'Microsoft.VisualBasic.CompilerServices.SiteDelegate1' does not exist in the reference but it does exist in the implementation.
-TypesMustExist : Type 'Microsoft.VisualBasic.CompilerServices.SiteDelegate2' does not exist in the reference but it does exist in the implementation.
-TypesMustExist : Type 'Microsoft.VisualBasic.CompilerServices.SiteDelegate3' does not exist in the reference but it does exist in the implementation.
-TypesMustExist : Type 'Microsoft.VisualBasic.CompilerServices.SiteDelegate4' does not exist in the reference but it does exist in the implementation.
-TypesMustExist : Type 'Microsoft.VisualBasic.CompilerServices.SiteDelegate5' does not exist in the reference but it does exist in the implementation.
-TypesMustExist : Type 'Microsoft.VisualBasic.CompilerServices.SiteDelegate6' does not exist in the reference but it does exist in the implementation.
-TypesMustExist : Type 'Microsoft.VisualBasic.CompilerServices.SiteDelegate7' does not exist in the reference but it does exist in the implementation.
-
-# C# doesn't permit setting the parameter name of the setter.
```
Comment (eerhardt) [inline src/libraries/Microsoft.VisualBasic.Core/src/MatchingRefApiCompatBaseline.txt:25]: Does this comment still add value? If so, it should be kept.

File: src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/Interaction.vb [hunk 8]
```diff
@@ -133,8 +134,13 @@ Namespace Microsoft.VisualBasic
             Return DirectCast(InvokeMethod("MsgBox", Prompt, Buttons, Title), MsgBoxResult)
         End Function
 
+        <UnconditionalSuppressMessage("ReflectionAnalsys", "IL2035:UnresolvedAssembly",
+                Justification:="This references Microsoft.VisualBasic.Forms assembly which is outside of the shared framework.")>
+        <UnconditionalSuppressMessage("ReflectionAnalsys", "IL2075:UnrecognizedReflectionPattern",
+                Justification:="DynamicDependency will preserve the required properties in order for the call to be safe.")>
+        <DynamicDependency(DynamicallyAccessedMemberTypes.PublicMethods, "Microsoft.VisualBasic._Interaction", "Microsoft.VisualBasic.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a")>
```
Comment (eerhardt) [inline src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/Interaction.vb:141]: I'm not sure this is necessary. If the assembly is there, the linker should be able to tell it needs to preserve public methods on this type.

cc @vitek-karas

File: src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/FileSystem.vb [hunk 9]
```diff
@@ -719,11 +725,12 @@ Namespace Microsoft.VisualBasic
             End Try
         End Sub
 
-        <ObsoleteAttribute("This member has been deprecated. Please use FilePutObject to write Object types, or coerce FileNumber and RecordNumber to Integer for writing non-Object types. https://go.microsoft.com/fwlink/?linkid=14202")>
```
Comment (eerhardt) [inline src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/FileSystem.vb:722]: Intentional to change the message here?

File: src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/Helpers/UnsafeNativeMethods.vb [hunk 10]
```diff
@@ -29,13 +30,19 @@ Namespace Microsoft.VisualBasic.CompilerServices
         End Function
 
         <DllImport("oleaut32", PreserveSig:=False, CharSet:=CharSet.Unicode, EntryPoint:="VarNumFromParseNum")>
+        <RequiresUnreferencedCode("Marshalling COM Objects is not trim safe.")>
+        <UnconditionalSuppressMessage("ReflectionAnalysis", "IL2050:COMMarshalling",
+            Justification:="RequiresUnreferencedCode attribute currently doesn't suppress IL2050. This should be removed once it does.")>
```
Comment (eerhardt) [inline src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/Helpers/UnsafeNativeMethods.vb:35]: Maybe add a link to the issue?

File: src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/FileIO/SpecialDirectories.vb [hunk 11]
```diff
@@ -155,8 +166,9 @@ Namespace Microsoft.VisualBasic.FileIO
             Return FileSystem.NormalizePath(Directory)
         End Function
 
+        <RequiresUnreferencedCode("Cannot statically analyze the passed in type.")>
         Private Shared Function GetWindowsFormsDirectory(typeName As String, propertyName As String) As String
```
Comment (eerhardt) [inline src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/FileIO/SpecialDirectories.vb:170]: Could we change this to have a DynamicallyAccessedMembers attribute on the `typeName As String`? And then pass the fully-qualified type in above. That way, when the assembly is present, the linker will see the full type name, and know to preserve properties on the type.

File: src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/VB6RandomFile.vb [hunk 12]
```diff
@@ -10,7 +11,7 @@ Imports Microsoft.VisualBasic.CompilerServices.Utils
 
 Namespace Microsoft.VisualBasic.CompilerServices
 
-    <System.ComponentModel.EditorBrowsableAttribute(System.ComponentModel.EditorBrowsableState.Never)> _
+    <System.ComponentModel.EditorBrowsableAttribute(System.ComponentModel.EditorBrowsableState.Never)>
```
Comment (eerhardt) [inline src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/VB6RandomFile.vb:14]: Intentional?

File: src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/VB6File.vb [hunk 13]
```diff
@@ -77,9 +78,9 @@ Namespace Microsoft.VisualBasic.CompilerServices
         [Single] = tagVT.VT_R4
         [Double] = tagVT.VT_R8
         [String] = tagVT.VT_BSTR
-        [ByteArray] = tagVT.VT_UI1 Or _
```
Comment (eerhardt) [inline src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/VB6File.vb:80]: Are these changes intentional?

File: src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/Symbols.vb [hunk 14]
```diff
@@ -916,6 +921,8 @@ Namespace Microsoft.VisualBasic.CompilerServices
             ' For a WinRT object, we want to treat members of it's collection interfaces as members of the object 
             ' itself. So GetMembers calls here to find the member in all the collection interfaces that this object 
             ' implements.
+            <UnconditionalSuppressMessage("ReflectionAnalysis", "IL2065:UnrecognizedReflectionPattern",
+                Justification:="_type is annotated with .All, so it's Interfaces will be annotated as well and it is safe to call GetMember on the Interfaces")>
```
Comment (eerhardt) [inline src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/Symbols.vb:925]: We should mark these with the issue, so we can clean it up once the linker is fixed.

[https://github.com/mono/linker/issues/1731](https://github.com/mono/linker/issues/1731)

File: src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/OverloadResolution.vb [hunk 15]
```diff
@@ -1690,6 +1701,7 @@ skipargument:
 
         'may not want Method as TargetProcedure - may instead want to pass the required information in separately.
         'this means that for the simple case of only one method we do not need to allocate a Method object.
+        <RequiresUnreferencedCode("Cannot statically analyze the paremetr types of the targetProcedure")>
```
Comment (eerhardt) [inline src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/OverloadResolution.vb:1704]: `paremetr` => `parameter`

File: src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/IDOBinder.vb [hunk 16]
```diff
@@ -276,6 +287,7 @@ Namespace Microsoft.VisualBasic.CompilerServices
 
         Private ReadOnly _ignoreReturn As Boolean
 
+        <RequiresUnreferencedCode("This subclass of SetIndexBinder is unsafe. Marking ctor unsafe in order to suppress warnings for overriden methods as unsafe.")>
```
Comment (cston) [inline src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/IDOBinder.vb:290]: >overriden [](http://example.com/codeflow?start=132&length=9)

"overridden" here and elsewhere?

Comment (cston) [inline src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/IDOBinder.vb:290]: >"This subclass of SetIndexBinder is unsafe. Marking ctor unsafe in order to suppress warnings for overriden methods as unsafe." [](http://example.com/codeflow?start=34&length=127)

This string is repeated a number of times. Should this be a constant?

File: src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/LateBinding.vb [hunk 17]
```diff
@@ -1014,7 +1061,14 @@ Namespace Microsoft.VisualBasic.CompilerServices
         End Function
 
         <DebuggerHiddenAttribute(), DebuggerStepThroughAttribute()>
-        Private Shared Function FastCall(ByVal o As Object, ByVal method As MethodBase, ByVal Parameters As ParameterInfo(), ByVal args As Object(), ByVal objType As Type, ByVal objIReflect As IReflect) As Object
+        Private Shared Function FastCall(
+                ByVal o As Object,
+                ByVal method As MethodBase,
+                ByVal Parameters As ParameterInfo(),
+                ByVal args As Object(),
+                ByVal objType As Type,
```
Comment (cston) [inline src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/LateBinding.vb:1069]: >ByVal objType As Type [](http://example.com/codeflow?start=16&length=21)

Should this parameter be annotated with `<DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)>`?

File: src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/IDOBinder.vb [hunk 18]
```diff
@@ -284,6 +296,8 @@ Namespace Microsoft.VisualBasic.CompilerServices
             _ignoreReturn = ignoreReturn
         End Sub
 
+        <UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode",
+            Justification:="The constructor of this subclass has been anotated.")>
```
Comment (cston) [inline src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/IDOBinder.vb:300]: >anotated [](http://example.com/codeflow?start=70&length=8)

"annotated" here and elsewhere?

File: src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/ConversionResolution.vb (hunk 4 above)
Comment (joperezr) [reply src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/ConversionResolution.vb:124]: While you are correct, the whole reason for suppressing this here is so that I don't have to add a bunch of 
```
#If DEBUG Then
<RequiresUnreferencedCode("bla")>
#End If
```

Would that be preferred? Or is your comment more about the justification message itself?

File: src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/IDOBinder.vb (hunk 18 above)
Comment (cston) [inline src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/IDOBinder.vb:300]: >"The constructor of this subclass has been anotated." [](http://example.com/codeflow?start=27&length=53)

This string is repeated a number of times. Should it be a constant?

File: src/libraries/Microsoft.VisualBasic.Core/src/MatchingRefApiCompatBaseline.txt (hunk 6 above)
Comment (joperezr) [reply src/libraries/Microsoft.VisualBasic.Core/src/MatchingRefApiCompatBaseline.txt:3]: Sort of, yeah. Not really related to this PR in particular but when adding the annotations on the src I started getting new APICompat warnings on since they didn't match the ref (as I hadn't updated that yet at that point) so I regenerated this file a few times until finally I regenerated it one last time after adding the changes back to the ref assembly again. Long story short, these baselines had gone way out of date since we don't regenerate them regularly, so this is just the result of regenerating them today. Of course I could remove this change from the PR and regenerate them outside of this effort if preferred.

File: src/libraries/Microsoft.VisualBasic.Core/src/MatchingRefApiCompatBaseline.txt (hunk 7 above)
Comment (joperezr) [reply src/libraries/Microsoft.VisualBasic.Core/src/MatchingRefApiCompatBaseline.txt:25]: Not really sure why this was here but the whole purpose of these files is for them to be auto generated, and we were supposed to regenerate them every now and then (which clearly we haven't done in a while) so I wouldn't personally revert this change as comment will be removed next regeneration anyway.

File: src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/FileSystem.vb (hunk 9 above)
Comment (joperezr) [reply src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/FileSystem.vb:722]: great catch!

File: src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/VB6RandomFile.vb (hunk 12 above)
Comment (joperezr) [reply src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/VB6RandomFile.vb:14]: Nope, VS keeps removing that for me though, had to re-add it back via notepad. Not sure what it is for or if it was just a typo, @cston any clues?

Comment (cston) [reply src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/VB6RandomFile.vb:14]: These are just legacy line continuation characters. If VS removes the continuation characters, it should be fine to remove them.

File: src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/LateBinding.vb (hunk 17 above)
Comment (joperezr) [reply src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/LateBinding.vb:1069]: Not really required as the actual type that is being reflected on is `objIReflect` when we pass it down to `InvokeMemberOnIReflect` so just annotating that seems to be enough.

File: src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/VB6RandomFile.vb (hunk 12 above)
Comment (joperezr) [reply src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/VB6RandomFile.vb]: oh, I saw your reply after I had reverted the changes back, I can take care of them again if you want me to @cston or I can leave them as is for now.

File: src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/FileIO/SpecialDirectories.vb (hunk 11 above)
Comment (joperezr) [reply src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/FileIO/SpecialDirectories.vb:164]: I remember I tried that, but then wasn't really sure if it would work since I was still getting the warning due to of course System.Windows.Forms not being present so I left it as RUC for now. I'll change it to be just annotated and suppress the warning for the missing assembly instead.

File: src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/VB6RandomFile.vb (hunk 12 above)
Comment (cston) [reply src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/VB6RandomFile.vb]: It's fine to leave the line continuation characters as is, or to remove them if VS removes them automatically, whichever is easier.

File: src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/ConversionResolution.vb (hunk 4 above)
Comment (eerhardt) [reply src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/ConversionResolution.vb:124]: I was thinking - why not just add the suppression to the `VerifyForLoopWidestType` method itself. Doesn't seem necessary to introduce a new method `VerifyForLoopWidestTypeInDebugMode` just to add the suppression.

File: src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/Symbols.vb [hunk 19]
```diff
@@ -932,6 +940,9 @@ Namespace Microsoft.VisualBasic.CompilerServices
                 Return result
             End Function
 
+            <UnconditionalSuppressMessage("ReflectionAnalysis", "IL2075:UnrecognizedReflectionPattern",
+                Justification:="_type is annotated with .All, so it's BaseType will be annotated as well and it is safe to call GetMember on the BaseType.
```
Comment (eerhardt) [inline src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/Symbols.vb:944]: I'm not sure I understand this message. Where is `BaseType` used?

File: src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/ConversionResolution.vb (hunk 4 above)
Comment (joperezr) [reply src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/ConversionResolution.vb:124]: Sure I can do that, the thinking was in case somebody else started calling this method should get the warning but you are correct that if they do it would be in DEBUG as well, so I like your suggestion, will go with that.

File: src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/Symbols.vb (hunk 19 above)
Comment (joperezr) [reply src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/Symbols.vb:944]: Apologies as it is not straight forward to figure this out. The variable classConstraint is created by calling `GetClassConstraint(_type)`, which if you see it's implementation, all it's doing is:

https://github.com/dotnet/runtime/blob/ac82799250fe42c8ba2941aef487aca94ecf04cc/src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/Symbols.vb#L522-L531

which basically means it is calling BaseType on `_type` and then the linker is warning because we are then calling GetMember into that and we haven't annotated the basetype, but because of the fact that `_type` itself is annotated with `DAM.All` then it's basetype and members should be preserved too.

**Review by eerhardt — APPROVED** [review]
LGTM. Thanks @joperezr!
